Fix KdocMethods#addKdocTemplate

KDocs added by `KDocMethods` rule are not properly indented by `IndentationRule`. This is probably because KDoc template is added as `LeafPsiElement`, while it should be a `CompositeElement`.
Example of `CompositeElement` construction can be found in KdocUtils: `ASTNode.insertTagBefore`.
Once this is fixed, one should try applying `KDocMethods` and `IndentationRule` simultaneously. Such case should also be added in #50 .

Turns out, KDocs are not so straightforwardly implemented as `CompositeElements`. When trying to modify `CompsiteElement(KDOC)`, e.g. `addChild`: 
```
Caused by: java.lang.ClassCastException: class org.jetbrains.kotlin.kdoc.lexer.KDocTokens$1 cannot be cast to class org.jetbrains.kotlin.KtNodeType
```

--- a/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocMethods.kt
+++ b/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocMethods.kt
@@ -19,32 +19,12 @@ import com.pinterest.ktlint.core.ast.ElementType.SAFE_ACCESS_EXPRESSION
 import com.pinterest.ktlint.core.ast.ElementType.THROW
 import com.pinterest.ktlint.core.ast.ElementType.TYPE_REFERENCE
 import com.pinterest.ktlint.core.ast.ElementType.WHEN_CONDITION_WITH_EXPRESSION
-import com.pinterest.ktlint.core.ast.ElementType.WHITE_SPACE
-import com.pinterest.ktlint.core.ast.prevSibling
 import org.cqfn.diktat.common.config.rules.RuleConfiguration
 import org.cqfn.diktat.common.config.rules.RulesConfig
 import org.cqfn.diktat.common.config.rules.getRuleConfig
-import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_TRIVIAL_KDOC_ON_FUNCTION
-import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_PARAM_TAG
-import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_RETURN_TAG
-import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_THROWS_TAG
-import org.cqfn.diktat.ruleset.constants.Warnings.MISSING_KDOC_ON_FUNCTION
+import org.cqfn.diktat.ruleset.constants.Warnings.*
 import org.cqfn.diktat.ruleset.rules.getDiktatConfigRules
-import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
-import org.cqfn.diktat.ruleset.utils.getBodyLines
-import org.cqfn.diktat.ruleset.utils.getFirstChildWithType
-import org.cqfn.diktat.ruleset.utils.getIdentifierName
-import org.cqfn.diktat.ruleset.utils.hasChildOfType
-import org.cqfn.diktat.ruleset.utils.hasKnownKDocTag
-import org.cqfn.diktat.ruleset.utils.hasTestAnnotation
-import org.cqfn.diktat.ruleset.utils.insertTagBefore
-import org.cqfn.diktat.ruleset.utils.isAccessibleOutside
-import org.cqfn.diktat.ruleset.utils.isGetterOrSetter
-import org.cqfn.diktat.ruleset.utils.isLocatedInTest
-import org.cqfn.diktat.ruleset.utils.isStandardMethod
-import org.cqfn.diktat.ruleset.utils.kDocTags
-import org.cqfn.diktat.ruleset.utils.parameterNames
-import org.cqfn.diktat.ruleset.utils.splitPathToDirs
+import org.cqfn.diktat.ruleset.utils.*
 import org.jetbrains.kotlin.com.intellij.lang.ASTNode
 import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LeafPsiElement
 import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.PsiWhiteSpaceImpl
@@ -217,18 +197,15 @@ class KdocMethods : Rule("kdoc-methods") {
                                 returnCheckFailed: Boolean
     ) {
         MISSING_KDOC_ON_FUNCTION.warnAndFix(configRules, emitWarn, isFixMode, name, node.startOffset) {
-            // fixme: indentation logic should be only inside IndentationRule
-            val indent = node.prevSibling { it.elementType == WHITE_SPACE }?.text
-                    ?.substringAfterLast("\n")?.count { it == ' ' } ?: 0
             val kDocTemplate = "/**\n" +
                     (missingParameters.joinToString("") { " * @param $it\n" } +
                             (if (returnCheckFailed) " * @return\n" else "") +
                             explicitlyThrownExceptions.joinToString("") { " * @throws $it\n" } +
                             " */\n"
-                            ).prependIndent(" ".repeat(indent))
-
-            // fixme could be added as proper CompositeElement
-            node.addChild(LeafPsiElement(KDOC, kDocTemplate), node.firstChildNode)
+                            )
+            val kdocNode = KotlinParser().createNode(kDocTemplate).findChildByType(KDOC)!!
+            node.appendNewlineMergingWhiteSpace(node.firstChildNode, node.firstChildNode)
+            node.addChild(kdocNode, node.firstChildNode)
         }
     }
 

--- a/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocMethods.kt
+++ b/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/KdocMethods.kt
@@ -22,7 +22,11 @@ import com.pinterest.ktlint.core.ast.ElementType.WHEN_CONDITION_WITH_EXPRESSION
 import org.cqfn.diktat.common.config.rules.RuleConfiguration
 import org.cqfn.diktat.common.config.rules.RulesConfig
 import org.cqfn.diktat.common.config.rules.getRuleConfig
-import org.cqfn.diktat.ruleset.constants.Warnings.*
+import org.cqfn.diktat.ruleset.constants.Warnings.MISSING_KDOC_ON_FUNCTION
+import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_PARAM_TAG
+import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_RETURN_TAG
+import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_WITHOUT_THROWS_TAG
+import org.cqfn.diktat.ruleset.constants.Warnings.KDOC_TRIVIAL_KDOC_ON_FUNCTION
 import org.cqfn.diktat.ruleset.rules.getDiktatConfigRules
 import org.cqfn.diktat.ruleset.utils.*
 import org.jetbrains.kotlin.com.intellij.lang.ASTNode

--- a/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/KdocMethodsFullExpected.kt
+++ b/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/KdocMethodsFullExpected.kt
@@ -17,20 +17,20 @@ class KdocMethodsFull {
 
     companion object {
         /**
-         * @param a
-         * @throws IllegalStateException
-         */
-        fun test3(a: Int) {
+ * @param a
+ * @throws IllegalStateException
+ */
+fun test3(a: Int) {
             throw IllegalStateException("Lorem ipsum")
         }
     }
 
     private class Nested {
         /**
-         * @param a
-         * @param b
-         * @return
-         */
-        fun test4(a: Int, b: Int): Int = 42
+ * @param a
+ * @param b
+ * @return
+ */
+fun test4(a: Int, b: Int): Int = 42
     }
 }

--- a/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/MissingKdocExpected.kt
+++ b/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/MissingKdocExpected.kt
@@ -2,12 +2,12 @@ package org.cqfn.diktat.kdoc.methods
 
 class Example {
     /**
-     * @param a
-     * @param b
-     * @return
-     * @throws IllegalStateException
-     */
-    fun addInts(a: Int, b: Int): Int {
+ * @param a
+ * @param b
+ * @return
+ * @throws IllegalStateException
+ */
+fun addInts(a: Int, b: Int): Int {
         if (false) throw IllegalStateException()
         return a + b
     }

--- a/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/MissingKdocWithModifiersExpected.kt
+++ b/diktat-rules/src/test/resources/test/paragraph2/kdoc/package/src/main/kotlin/org/cqfn/diktat/kdoc/methods/MissingKdocWithModifiersExpected.kt
@@ -2,12 +2,12 @@ package test.paragraph2.kdoc
 
 class Example {
     /**
-     * @param a
-     * @param b
-     * @return
-     * @throws IllegalStateException
-     */
-    @Deprecated("Use something else")
+ * @param a
+ * @param b
+ * @return
+ * @throws IllegalStateException
+ */
+@Deprecated("Use something else")
     internal fun addInts(a: Int, b: Int): Int {
         if (false) throw IllegalStateException()
         return a + b

